numba/_dispatcher.cpp

```diff
@@ -678,10 +710,158 @@ Dispatcher_call(Dispatcher *self, PyObject *args, PyObject *kws)
+        if (tys[i] == -1) {
+            if (self->can_fallback){
+                /* We will clear the exception if fallback is allowed. */
+                PyErr_Clear();
+            } else {
+                goto CLEANUP;
+            }
+        }
+    }
+
+    /* We only allow unsafe conversions if compilation of new specializations
+       has been disabled. */
+    cfunc = self->resolve(tys, matches, !self->can_compile,
+                          exact_match_required);
+
+    if (matches == 0 && !self->can_compile) {
+        /*
+         * If we can't compile a new specialization, look for
+         * matching signatures for which conversions haven't been
+         * registered on the C++ TypeManager.
+         */
+        int res = search_new_conversions((PyObject *) self, args, kws);
+        if (res < 0) {
+            retval = NULL;
+            goto CLEANUP;
+        }
+        if (res > 0) {
+            /* Retry with the newly registered conversions */
+            cfunc = self->resolve(tys, matches, !self->can_compile,
+                                  exact_match_required);
+        }
+    }
+
+    if (matches == 1) {
+        /* Definition is found */
+        retval = cfunc;
+        Py_INCREF(retval);
+    } else if (matches == 0) {
+        /* No matching definition */
+        if (self->can_compile) {
+            retval = cuda_compile_only(self, args, kws, locals);
+        } else if (self->fallbackdef) {
+            /* Have object fallback */
+            retval = call_cfunc(self, self->fallbackdef, args, kws, locals);
+        } else {
+            /* Raise TypeError */
+            explain_matching_error((PyObject *) self, args, kws);
+            retval = NULL;
+        }
+    } else if (self->can_compile) {
+        /* Ambiguous, but are allowed to compile */
+        retval = cuda_compile_only(self, args, kws, locals);
+    } else {
+        /* Ambiguous */
+        explain_ambiguous((PyObject *) self, args, kws);
+        retval = NULL;
+    }
+
+CLEANUP:
+    if (tys != prealloc)
+        delete[] tys;
+    Py_DECREF(args);
+
+    return retval;
+}
+
+static int
+import_devicearray(void)
+{
+    PyObject *devicearray = PyImport_ImportModule("numba._devicearray");
+    PyObject *c_api = NULL;
+
+    if (devicearray == NULL) {
+        return -1;
+    }
+
+    c_api = PyObject_GetAttrString(devicearray, "_DEVICEARRAY_API");
+    Py_DECREF(devicearray);
+    if (c_api == NULL) {
+        PyErr_SetString(PyExc_AttributeError, "_DEVICEARRAY_API not found");
+        return -1;
+    }
+
+    if (!PyCapsule_CheckExact(c_api)) {
+        PyErr_SetString(PyExc_RuntimeError, "_DEVICEARRAY_API is not PyCapsule object");
+        Py_DECREF(c_api);
+        return -1;
+    }
+
+    DeviceArray_API = (void**)PyCapsule_GetPointer(c_api, "_devicearray.c_api");
+    Py_DECREF(c_api);
+    if (DeviceArray_API == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "_DEVICEARRAY_API is NULL pointer");
+        return -1;
+    }
```

Comment:
Much of this code can be reduced if `PyImport_ImportModule` is used:
```suggestion
    if (devicearray == NULL) {
        return -1;
    }
    Py_DECREF(devicearray);

    DeviceArray_API = (void**)PyCapsule_Import("numba._devicearray._DEVICEARRAY_API", 0);
    if (DeviceArray_API == NULL) {
        return -1;
    }
```
```